ddrutility.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.44.1.
.TH DDRUTILITY "1" "October 2022" "ddrutility 2.8.1 20221013" "User Commands"
.SH NAME
ddrutility \- manual page for ddrutility 2.8.1 20221013
.SH SYNOPSIS
.B ddrutility
[\fIoptions\fR]
.SH DESCRIPTION
ddrutility \- Tool set to aid in data recovery
.PP
Ddrutility is a toolset designed to help aid in data recovery using GNU
ddrescue. It contains the following programs:
.PP
Ddru_findbad \- Utility for finding what files are related to bad sectors. For
more info on ddru_findbad, try 'ddru_findbad \fB\-\-help\fR'.
.PP
Ddru_ntfsbitmap \- Utility to extract the bitmap file from a NTFS partition, and
then create a domain log file to use with ddrescue that would only recover the
used portion of the partition. For more info on ddru_ntfsbitmap, try
\&'ddru_ntfsbitmap \fB\-\-help\fR'.
.PP
Ddru_ntfsfindbad \- Utility for NTFS partitions to find which files are related
to bad sectors in a ddrescue log file. You should use this for NTFS partitions
in place of the original ddru_findbad. For more info on ddru_ntfsfindbad, try
\&'ddru_ntfsfindbad \fB\-\-help\fR'.
.PP
Ddru_diskutility \- Advanced LINUX ONLY utility that can perform several
different functions on a disk, using some direct pass\-through disk commands. For
more info on ddru_diskutility, try 'ddru_diskutility \fB\-\-help\fR'.
.SS "OPTIONS:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
Show version and exit
.SH "REPORTING BUGS"
Report bugs to maximus57@hotmail.com
.br
Or better yet, report bugs at the project home page
.br
Ddrutility homepage: http://sourceforge.net/projects/ddrutility/
.SH COPYRIGHT
Copyright \(co 2016 Scott Dwyer.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B ddrutility
is maintained as a Texinfo manual.  If the
.B info
and
.B ddrutility
programs are properly installed at your site, the command
.IP
.B info ddrutility
.PP
should give you access to the complete manual.


ChangeLog
version 2.8.1 released October 13, 2022
  * fix too small char buffer in ddru_ntfsbitmap.c
  * change makefile to make compile successful on newer gcc

version 2.8 released November 23, 2016
  * No longer uses automake
  * Uses regular make file
  * No program changes were made

version 2.7 released January 11, 2015
  * Ddru_ntfsfindbad 1.5 released:
    + Fixed possible program crash if partition boot sector error
    + Better partition boot sector error output
  * Ddru_ntfsbitmap 1.5 released:
    + Fixed possible program crash if partition boot sector error
    + Better partition boot sector error output

version 2.6 released October 15 2014
  * Changes have been made for compiling compatibility:
    + Some unneeded items removed from configure.ac
    + Added lib check for iconv
  * Some improvements have been made to the documentation:
    + Added examples to the --mftdomain option of ntfs_bitmap
    + Updated info about ddru_findbad being slow
  * Ddru_findbad 1.11 released:
    + No longer relies on bash
    + Fixed a bug dealing with bad ntfscluster results
    + Images are now accessed as read only
  * Ddru_ntfsfindbad 1.4 released:
    + Fixed potential memory bug with name conversions
    + Fixed iconv BOM issue
    + Fixed a bug with mft data run length
    + Fixed issue with current postition in logfile
  * Ddru_ntfsbitmap 1.4 released:
    + Fixed potential memory bug with name conversions
    + Fixed iconv BOM issue
  * Ddru_diskutility 1.3 released:
    + Initial release

version 2.5 released June 8 2014
  * Some code restructuring has been done:
  * Ddru_ntfsbitmap 1.3 released:
    + Now able to process non 512 byte sectors
    + Can process some partial MFT errors
    + Fixed a bug in MFT domain logfile creation
    + Now able to process older NTFS version
  * Ddru_ntfsfindbad 1.3 released:
    + Now able to process non 512 byte sectors
    + Can process some partial MFT errors
    + Now able to process older NTFS version

version 2.4 released May 15 2014
  * Ddru_ntfsbitmap 1.2 released:
    + Added option --restart to delete all files at start
    + Fixed a bug in domain logfile creation
  * Ddru_ntfsfindbad 1.2 released:
    + Fixed a bug where 0 byte errors were reported
    + Fixed a bug in the debug output

version 2.3
  * Ddru_ntfsfindbad 1.1 released:
    + Fixed bug where filename is output in DOS format
    + Now converts NTFS filenames from UTF-16 to UTF-8
    + Added option --encoding to choose output encoding
    + Added option --noconvert to turn off iconv conversion

version 2.2
  * Ddru_ntfsbitmap 1.1 released:
    + Added option to bridge small gaps of unused space
    + Boot sector file now deleted if not NTFS
    + Improved handling of ddrescue return codes
  * Ddru_findbad 1.10 released:
    + Added option to manually control partition types

version 2.1
  * Ddru_findbad 1.9 released:
    + Now uses fdisk if gfdisk is not available
  * Ddru_ntfsfindbad 1.0 released:
    + Initial release

version 2.0
  * Documentation updates
    + Added requirements to the top of the readme file
  * Ddru_ntfsbitmap 1.0 released:
    + Made some changes to the code for portability
  * Ddru_findbad 1.8 released:
    + Fixed a bug that could cause an error processing the log file
    + Now displays the logfile lines as it processes them

version 1.99-beta2
  * Documentation updates
    + New section added "Important Notes"
    + Ddru_ntfsbitmap documentation improvements
  * Changes and improvements to ddru_ntfsbitmap
    + New option added to ddru_ntfsbitamp, "--mftdomain"
    + Output report now contains domain logfile info
    + Fixed bug where domain logfile did not include all drive space
    + Output file __mftbitmap renamed __mftshort and includes more

version 1.99-beta1
  * Complete change in packaging - now uses make
  * Ddrutility is now a collection of programs
  * Original ddrutility renamed to ddru_findbad
  * Ddru_findbad 1.7 released:
    + No longer counts non-tried "?" areas
    + Quits if nothing bad found in logfile instead of producing errors
    + Removed full documentation from --help and man
    + Full documentation now located in info ddrutility
    + Added examples to documentation
  * Ddru_ntfsbitmap 0.99 beta1 released"
    + Initial release

version 1.6
  * Fixed bug where ddrescue logfile was not processed

version 1.5
  * No longer uses ddrescuelog
  * Improved error handling
    + Improved some error and warning messages for easier understanding
    + Eliminated some conditions that caused errors
    + Most error messages go to the debug file
  * Uses single base name for output files
    + Reconstructed output files for different uses
  * Improved processing speed
    + Only checks once per cluster/block instead of every sector
  * Shows total run time when done
  * Sector file can contain ranges
  * Added option to produce extra types of output files
  * Fixed command input problem that caused infinite loop
  * Added quick processing option
  * Added special quick ntfs processing option
  * Added version number to output files
  * Many more small changes and improvements too numerous to list

version 1.4
  * Added MAN page
    + Made improvements to --help output
  * Added support for HFS+ file system
  * Added support for GPT and APT partition tables
  * Improved some of the code
    + Improved how it handles loop devices
    + Improved handling of FAT partitions
    + Changed names of temp files so less likely to be existing file
    + Target, logfile, and output file names can now contain spaces
    + Added option to change sector size
    + Added option to include more info in long results
    + Changed command line arguments
  * Fixed bug so Linux partitions get correct block size
  * Major code reconstruction to make it more modular

version 1.3
  * Initial release



ddrutility.c
/*
Copyright (C) 2016 Scott Dwyer

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <stdbool.h>

#include "ddrutility_help.h"

char *title = "ddrutility";
char *version_number = "2.8.1 20221013";
int copyright_year = 2022;

void help(void);
void version(void);

int main (int argc, char **argv)
{

  // begin processing command line arguments
  int command_line_argument;
  bool command_line_error = false;	// initialize error to false

  while (1)
  {
    static struct option long_options[] =
    {
      {"help",    	no_argument,	0, 'h'},
      {"version", 	no_argument,	0, 'v'},
      {0, 0, 0, 0}
    };
    // getopt_long stores the option index here.
    int option_index = 0;

    command_line_argument = getopt_long (argc, argv, "hv",
					 long_options, &option_index);

    // Detect the end of the options.
    if (command_line_argument == -1)
      break;

    switch (command_line_argument)
    {
      case 'h':
	help();
	exit(0);

      case 'v':
	version();
	exit(0);

      case '?':
	// getopt_long already printed an error message.
	command_line_error = true;
	break;

      default:
	puts ("aborting");
	abort ();
    }
  }

  // exit on command line error
  if (command_line_error)
  {
    puts ("ERROR: command line input error");
    exit (3);
  }
  // end of command line processing
  printf ("%s - data recovery utility tool set\n", title);
  printf ("try '%s --help' for more information\n", title);
  return(0);
}


// function to display help
void help(void)
{
  unsigned int i;
  for (i = 0; i < ddrutility_help_txt_len; i++)
  {
    printf ("%c", ddrutility_help_txt[i]);
  }
}


// function to display version
void version(void)
{
  printf ("%s %s\n", title, version_number);
  printf ("Copyright (C) %d Scott Dwyer.\n", copyright_year);
  printf ("License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n");
  printf ("This is free software: you are free to change and redistribute it.\n");
  printf ("There is NO WARRANTY, to the extent permitted by law.\n");
}